### ISSUE TITLE
Response channel for stream X not found error after cancelling HTTP request context

### ISSUE THREAD

**Issue by rayvbr** [lucas-clemente/quic-go#1455]
It seems quic-go's internal state gets broken when one or more HTTP requests are cancelled mid-stream. When this happens, a significant part of any existing and future HTTP requests over the same connection will fail with `InvalidHeadersStreamData: response channel for stream X not found`

To reproduce:

- Create N http clients through:
```
httpClients = make([]http.Client, N)
for i := 0; i < N; i++ {
  transport := &h2quic.RoundTripper{
				QuicConfig: &quic.Config{
					KeepAlive:                   true,
					RequestConnectionIDOmission: false,
				},
				DisableCompression: true,
			}
  httpClients[i] = http.Client{Transport: transport}
}
```

- Schedule M HTTP requests on the N http clients (where M > N). Use `http.NewRequest("GET", url, nil).WithContext(ctx)` to allow cancellation of the HTTP requests.
- Randomly cancel a few HTTP requests by calling ctx.Cancel()

A very significant percentage of the HTTP requests will fail with the following error: `InvalidHeadersStreamData: response channel for stream X not found`

**Comment by rayvbr** [comment]
Note that after analysing lots of logs, I noticed that irrespective of how many http requests I performed over the same connection, the InvalidHeadersStreamData error always reported the same stream IDs to not be found. In other words, it seems the stream counter no longer counts up after this error happens for the first time.

**Comment by marten-seemann** [comment]
Thanks for reporting! 👍 

You don't even need N clients, using one client is enough to reproduce that error. All requests after the canceled request fail.
The problem seems to be that if the request was already sent out, we need to read the header frame of the response, and discard it.

**Comment by rayvbr** [comment]
Yes, you're absolutely right. Thanks!

**Comment by marten-seemann** [comment]
Hello @rayvbr, I already have a fix for a couple of weeks, but I'm struggling with the test cases. I hope I'll find the time to fix everything soon.

**Comment by rayvbr** [comment]
Thanks @marten-seemann! If there's anything I can do to help, let me know

**Comment by marten-seemann** [comment]
Closed by #1852.
